src/ripple/legacy/0.27/CreateOffer.cpp

```diff
@@ -0,0 +1,631 @@
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/legacy/0.27/book/OfferStream.h>
+#include <ripple/legacy/0.27/book/Taker.h>
+#include <ripple/legacy/0.27/book/Types.h>
+#include <ripple/legacy/0.27/book/Amounts.h>
+#include <ripple/legacy/0.27/CreateOffer.h>
+#include <ripple/basics/Log.h>
+#include <ripple/core/Config.h>
+#include <ripple/json/to_string.h>
+#include <beast/cxx14/memory.h>
+
+namespace ripple {
+
+namespace legacy {
+
+TER
+CreateOffer::checkAcceptAsset(IssueRef issue) const
+{
+    /* Only valid for custom currencies */
+    assert (!isXRP (issue.currency));
+
+    SLE::pointer const issuerAccount = mEngine->entryCache (
+        ltACCOUNT_ROOT, getAccountRootIndex (issue.account));
+
+    if (!issuerAccount)
+    {
+        if (m_journal.warning) m_journal.warning <<
+            "delay: can't receive IOUs from non-existent issuer: " <<
+            to_string (issue.account);
+
+        return (mParams & tapRETRY)
+            ? terNO_ACCOUNT
+            : tecNO_ISSUER;
+    }
+
+    if (issuerAccount->getFieldU32 (sfFlags) & lsfRequireAuth)
+    {
+        SLE::pointer const trustLine (mEngine->entryCache (
+            ltRIPPLE_STATE, getRippleStateIndex (
+                mTxnAccountID, issue.account, issue.currency)));
+
+        if (!trustLine)
+        {
+            return (mParams & tapRETRY)
+                ? terNO_LINE
+                : tecNO_LINE;
+        }
+
+        // Entries have a canonical representation, determined by a
+        // lexicographical "greater than" comparison employing strict weak
+        // ordering. Determine which entry we need to access.
+        bool const canonical_gt (mTxnAccountID > issue.account);
+
+        bool const is_authorized (trustLine->getFieldU32 (sfFlags) &
+            (canonical_gt ? lsfLowAuth : lsfHighAuth));
+
+        if (!is_authorized)
+        {
+            if (m_journal.debug) m_journal.debug <<
+                "delay: can't receive IOUs from issuer without auth.";
+
+            return (mParams & tapRETRY)
+                ? terNO_AUTH
+                : tecNO_AUTH;
+        }
+    }
+
+    return tesSUCCESS;
+}
+
+std::pair<TER, legacy::core::Amounts>
+CreateOffer::crossOffers (legacy::core::LedgerView& view,
+    legacy::core::Amounts const& taker_amount)
+{
+    legacy::core::Taker::Options const options (mTxn.getFlags());
```

Comment:
This block of code might well be more readable using `auto`, but getting rid of the unneeded `legacy::` would also help.